ext/js/dictionary/dictionary-data-util.js

```diff
@@ -288,26 +321,43 @@ export class DictionaryDataUtil {
 
     /**
      * @param {string} reading
-     * @param {number} position
-     * @param {number[]} nasalPositions
-     * @param {number[]} devoicePositions
-     * @param {import('dictionary').Tag[]} tags
+     * @param {import('dictionary').Pronunciation} pronunciation
      * @param {import('dictionary-data-util').GroupedPronunciationInternal[]} groupedPronunciationList
      * @returns {?import('dictionary-data-util').GroupedPronunciationInternal}
      */
-    static _findExistingGroupedPronunciation(reading, position, nasalPositions, devoicePositions, tags, groupedPronunciationList) {
-        for (const pitchInfo of groupedPronunciationList) {
-            if (
-                pitchInfo.reading === reading &&
-                pitchInfo.position === position &&
-                this._areArraysEqual(pitchInfo.nasalPositions, nasalPositions) &&
-                this._areArraysEqual(pitchInfo.devoicePositions, devoicePositions) &&
-                this._areTagListsEqual(pitchInfo.tags, tags)
-            ) {
-                return pitchInfo;
+    static _findExistingGroupedPronunciation(reading, pronunciation, groupedPronunciationList) {
+        const existingGroupedPronunciation = groupedPronunciationList.find((groupedPronunciation) => {
+            return groupedPronunciation.reading === reading && this._arePronunciationsEquivalent(groupedPronunciation, pronunciation);
+        });
+
+        return existingGroupedPronunciation || null;
+    }
+
+    /**
+     * @param {import('dictionary-data-util').GroupedPronunciationInternal} groupedPronunciation
+     * @param {import('dictionary').Pronunciation} pronunciation
+     * @returns {boolean}
+     */
+    static _arePronunciationsEquivalent(groupedPronunciation, pronunciation) {
+        if (!this._areTagListsEqual(groupedPronunciation.tags, pronunciation.tags)) {
+            return false;
+        }
+
+        if (pronunciation.type === 'pitch-accent' && groupedPronunciation.type === 'pitch-accent') {
```

Comment:
This can be simplified. Put the type check at the start of the function; that check is faster, and if those aren't the same, nothing else is.

`if (p.type !== gp.type) { return false; }`

Then you can `switch` on type.

Reply:
I tried that first, but didn't like the redundant type check in each case to get it to calm down about the types:

```
static _arePronunciationsEquivalent({pronunciation: pronunciation1}, pronunciation2) {
        if (pronunciation1.type !== pronunciation2.type) {
            return false;
        }
        if (!this._areTagListsEqual(pronunciation1.tags, pronunciation2.tags)) {
            return false;
        }
        switch (pronunciation1.type) {
            case 'pitch-accent':
                if (
                    pronunciation2.type === 'pitch-accent' &&
                    (pronunciation1.position !== pronunciation2.position ||
                     !this._areArraysEqual(pronunciation1.nasalPositions, pronunciation2.nasalPositions) ||
                     !this._areArraysEqual(pronunciation1.devoicePositions, pronunciation2.devoicePositions))
                ) {
                    return false;
                }
                break;
            case 'phonetic-transcription':
                if (
                    pronunciation2.type === 'phonetic-transcription' &&
                    pronunciation1.ipa !== pronunciation2.ipa
                ) {
                    return false;
                }
                break;
        }

        return true;
    }
```

Is that still preferable?

Reply:
Oh that's annoying that TypeScript can't infer that. Here's the simplest it can probably be made:

```js
static _arePronunciationsEquivalent({pronunciation: pronunciation1}, pronunciation2) {
    if (
        pronunciation1.type !== pronunciation2.type ||
        !this._areTagListsEqual(pronunciation1.tags, pronunciation2.tags)
    ) {
        return false;
    }
    switch (pronunciation1.type) {
        case 'pitch-accent':
        {
            // This cast is valid based on the type check at the start of the function.
            const pitchAccent2 = /** @type {import('dictionary').PitchAccent} */ (pronunciation2);
            return (
                pronunciation1.position === pitchAccent2.position &&
                this._areArraysEqual(pronunciation1.nasalPositions, pitchAccent2.nasalPositions) &&
                this._areArraysEqual(pronunciation1.devoicePositions, pitchAccent2.devoicePositions)
            );
        }
        case 'phonetic-transcription':
        {
            // This cast is valid based on the type check at the start of the function.
            const phoneticTranscription2 = /** @type {import('dictionary').PhoneticTranscription} */ (pronunciation2);
            return pronunciation1.ipa === phoneticTranscription2.ipa;
        }
    }
    return true;
}
```